### PR TITLE
Updated src/defaults/schema.json

### PR DESCRIPTION
**Pull request by rudrapsc**
edited schema, now it will show validation error if in the preprocessor the name is not given , options are not given and if the name of the plugin happen to be CropOnMarkers , then relative path becomes a required field and will show error if not given.

removed relative path from the template.json of sample4
image

it showed this error
image

### CODE REVIEW

File: src/defaults/schema.json [hunk 1]
```diff
@@ -27,9 +27,73 @@
       "description": "Custom configuration values to use in the template's directory",
       "type": "array",
       "items": {
-        "type": "object"
-      }
-    },
+        "type": "object",
+        "properties": {
+        "name": {
+          "enum": ["CropOnMarkers","CropPage","FeatureBasedAlignment","ReadBarcode","GaussianBlur","Levels","MedianBlur"],
```
Comment (Udayraj123) [inline src/defaults/schema.json:33]: format the json - spaces between items

**Comment by rudrapsc** [comment]
image

File: src/defaults/schema.json [hunk 2]
```diff
@@ -27,9 +27,164 @@
...
+                "options": {
+                  "type": "object",
+                  "propertyNames": {
+                  "enum": ["relativePath","min_matching_threshold","max_matching_variation","marker_rescale_range",
+                    "marker_rescale_steps","apply_erode_subtract","sheetToMarkerWidthRatio"]
+                  },
+                  "properties": {
+                    "relativePath": {
+                      "type": "string"
+                    }
```
Comment (Udayraj123) [inline src/defaults/schema.json:60]: Add types for min_matching_threshold and others too (float, int, etc)

File: src/defaults/schema.json [hunk 3]
```diff
@@ -27,9 +27,164 @@
...
+                   "const": "CropPage"
+                 }
+               }
+             },
+             "then": {
+               "properties": {
+                 "options": {
+                   "type": "object",
+                   "propertyNames": {
+                  "enum": ["morphKernel"]
```
Comment (Udayraj123) [inline src/defaults/schema.json:175]: also define type for morphKernel as a 2-sized array

File: src/defaults/schema.json [hunk 4]
```diff
@@ -27,9 +27,164 @@
...
+                  "const": "CropOnMarkers"
+                }
+              }
+            },
+            "then": {
+              "properties": {
+                "options": {
+                  "type": "object",
+                  "propertyNames": {
+                  "enum": ["relativePath","min_matching_threshold","max_matching_variation","marker_rescale_range",
```
Comment (Udayraj123) [inline src/defaults/schema.json:54]: Do the space formatting changes too (use json beautifier)

File: src/defaults/schema.json [hunk 5]
```diff
@@ -27,7 +27,273 @@
...
+                      "type": "array",
+                      "prefixItems": [
+                        {
+                          "type": "integer"
+                        },
+                        {
+                          "type": "integer"
+                        }
+                      ],
+                      "maxItems": 2
```
Comment (Udayraj123) [inline src/defaults/schema.json:92]:
```suggestion
                      "minItems": 2,
                      "maxItems": 2
```

File: src/defaults/schema.json [hunk 6]
```diff
@@ -27,7 +27,273 @@
...
+                    },
+                    "marker_rescale_steps": {
+                      "type": "integer"
+                    },
+                    "apply_erode_subtract": {
+                      "type": "integer"
+                    },
+                    "sheetToMarkerWidthRatio": {
+                      "type": "number"
```
Comment (Udayraj123) [inline src/defaults/schema.json:101]: okay, so "number" allows floats + integers both, cool

File: src/defaults/schema.json [hunk 7]
```diff
@@ -27,7 +27,273 @@
...
+                  "type": "object",
+                  "propertyNames": {
+                    "enum": [
+                      "morphKernel"
+                    ]
+                  },
+                  "properties": {
+                    "morphKernel": {
+                      "type": "array",
+                      "prefixItems": [
```
Comment (Udayraj123) [inline src/defaults/schema.json:277]: Why do we need prefixItems if type of both items is integer?

File: src/defaults/schema.json [hunk 8]
```diff
@@ -27,7 +27,273 @@
...
+                        {
+                          "type": "integer"
+                        },
+                        {
+                          "type": "integer"
+                        }
+                      ],
+                      "maxItems": 2
```
Comment (Udayraj123) [inline src/defaults/schema.json:285]: add minItems = 2 as well

File: src/defaults/schema.json [hunk 9]
```diff
@@ -27,7 +27,273 @@
...
+                    }
+                  }
+                }
+              }
+            }
+          }
+        ],
+        "required": [
```
Comment (Udayraj123) [inline src/defaults/schema.json:293]: move this key to the top before "allOf"

**Review by Udayraj123 — CHANGES_REQUESTED** [review]
Suggested few more changes. Also add 1-2 more screenshots of invalid jsons

**Comment by rudrapsc** [comment]
image

this screenshot shows three invalid json ,one with array length greater than the max , one with array length shorter than given min and one with garbage options.

**Comment by Udayraj123** [comment]
Okay, the key column seems not specific to the error path

**Comment by Udayraj123** [comment]
Can you make the following change in `src/utils/file.py` and share the output with me?
```python
key, validator, msg = error.path, error.validator, error.message
# print the error paths
print(error.path)
print(error.absolute_path)
```

ref: https://python-jsonschema.readthedocs.io/en/latest/errors/

**Comment by rudrapsc** [comment]
> Can you make the following change in `src/utils/file.py` and share the output with me?
> 
> ```python
> key, validator, msg = error.path, error.validator, error.message
> # print the error paths
> print(error.path)
> print(error.absolute_path)
> ```
> 
> ref: https://python-jsonschema.readthedocs.io/en/latest/errors/

image

this is the output i received

**Comment by Udayraj123** [comment]
image
Need to change sigmax in builtins.py

**Comment by Udayraj123** [comment]
Changing morphKernel to have 1.0 as one of the values gives this error -
```
TypeError: Can't parse 'ksize'. Sequence item with index 0 has a wrong type

```

**Pull request merged**